### PR TITLE
Replace panic! with internal-error! in ast crate

### PR DESCRIPTION
**Pull request by lucacervello**
Hi,

I replaced `panic!` with `internal_error!`.
I started with only `panic!` and only inside the `ast` crate to keep this PR small.
Related to https://github.com/roc-lang/roc/issues/2046

### CODE REVIEW

File: crates/ast/src/canonicalization/canonicalize.rs [hunk 1]
```diff
@@ -188,7 +189,7 @@ fn canonicalize_field<'a>(
         }
 
         Malformed(_string) => {
-            panic!("TODO canonicalize malformed record field");
+            internal_error!("TODO canonicalize malformed record field");
```
Comment (bhansconnect) [inline crates/ast/src/canonicalization/canonicalize.rs:192]: Can we making things that say `TODO` into `todo!`?

File: crates/ast/src/constrain.rs [hunk 2]
```diff
@@ -2064,7 +2065,7 @@ pub mod test_constrain {
 
                 assert_eq!(actual_str, expected_str);
             }
-            Err(e) => panic!("syntax error {:?}", e),
+            Err(e) => internal_error!("syntax error {:?}", e),
```
Comment (bhansconnect) [inline crates/ast/src/constrain.rs:2068]: I am not very fimilare with this code, but I wonder if this is a `user_error!` that just doesn't have a pretty message yet.

File: crates/ast/src/lang/core/def/def.rs [hunk 3]
```diff
@@ -574,7 +574,7 @@ fn canonicalize_pending_def<'a>(
                                     ..
                                 } => {
                                     if arguments.len() != type_arguments.len() {
-                                        panic!("argument number mismatch");
+                                        internal_error!("argument number mismatch");
```
Comment (bhansconnect) [inline crates/ast/src/lang/core/def/def.rs:577]: Same with this and the errors below in this file. They might really be `user_error!`

File: crates/ast/src/solve_type.rs [hunk 4]
```diff
@@ -113,15 +113,15 @@ impl Pools {
     }
 
     pub fn get_mut(&mut self, rank: Rank) -> &mut Vec<Variable> {
-        self.0
-            .get_mut(rank.into_usize())
-            .unwrap_or_else(|| panic!("Compiler bug: could not find pool at rank {}", rank))
+        self.0.get_mut(rank.into_usize()).unwrap_or_else(|| {
```
Comment (bhansconnect) [inline crates/ast/src/solve_type.rs:116]: nit: remove the `Compiler bug:` prefix here and below

File: crates/ast/src/module.rs [hunk 5]
```diff
@@ -20,14 +21,16 @@ pub fn load_module(src_file: &Path, threading: Threading) -> LoadedModule {
     match loaded {
         Ok(x) => x,
         Err(roc_load::LoadingProblem::FormattedReport(report)) => {
-            panic!(
+            internal_error!(
```
Comment (bhansconnect) [inline crates/ast/src/module.rs:24]: Also, here, not sure if this should technically be a `user_error!`

**Review by bhansconnect — COMMENTED** [review]
@ayazhafiz or @folkertdev, could you take a look at this? Some of these cases look like the might be `user_error!` instead of `internal_error!`. I just don't know this code well enough.

**Comment by bhansconnect** [comment]
Thanks so much for doing this! These are the small steps to eventually getting a robust and fuzzable compiler. Makes it clear to the end users when they should to file bugs.

File: crates/ast/src/constrain.rs (hunk 2 above)
Comment (ayazhafiz) [reply crates/ast/src/constrain.rs:2068]: I think using `internal_error` is still okay here, things that don't exit well to the user are internal errors to the compiler.

**Review by ayazhafiz — COMMENTED** [review]
LGTM after @bhansconnect 's comments are addressed

File: crates/ast/src/solve_type.rs [hunk 6]
```diff
@@ -115,13 +115,13 @@ impl Pools {
...
-            .unwrap_or_else(|| panic!("Compiler bug: could not find pool at rank {}", rank))
+            .unwrap_or_else(|| internal_error!("could not find pool at rank {}", rank))
     }
 
     pub fn get(&self, rank: Rank) -> &Vec<Variable> {
-        self.0
-            .get(rank.into_usize())
-            .unwrap_or_else(|| panic!("Compiler bug: could not find pool at rank {}", rank))
+        self.0.get(rank.into_usize()).unwrap_or_else(|| {
+            internal_error!("Compiler bug: could not find pool at rank {}", rank)
```
Comment (bhansconnect) [inline crates/ast/src/solve_type.rs:123]: nit: remove `Compiler bug:` here as well.

File: crates/ast/src/canonicalization/canonicalize.rs [hunk 7]
```diff
@@ -188,7 +189,7 @@ fn canonicalize_field<'a>(
         }
 
         Malformed(_string) => {
-            panic!("TODO canonicalize malformed record field");
+            internal_error!("todo! canonicalize malformed record field");
```
[the remark this answers is not in the document]
Comment (lucacervello) [reply crates/ast/src/canonicalization/canonicalize.rs:192]: Oh, yes of course.

**Review by bhansconnect — APPROVED** [review]
Sorry for being so slow about this. The last few weeks have been quite hectic for me. Looks great!

**Pull request closed without merging**